warehouse/oidc/models/google.py

```diff
@@ -68,20 +69,20 @@ class GooglePublisherMixin:
 
     __unchecked_claims__ = {"azp", "google"}
 
-    @staticmethod
-    def __lookup_all__(klass, signed_claims: SignedClaims) -> Query | None:
-        return Query(klass).filter_by(
-            email=signed_claims["email"], sub=signed_claims["sub"]
-        )
+    @classmethod
+    def lookup_by_claims(cls, session, signed_claims: SignedClaims) -> Self:
+        query: Query = Query(cls).filter_by(email=signed_claims["email"])
+        publishers = query.with_session(session).all()
 
-    @staticmethod
-    def __lookup_no_sub__(klass, signed_claims: SignedClaims) -> Query | None:
-        return Query(klass).filter_by(email=signed_claims["email"], sub="")
+        if sub := signed_claims.get("sub"):
+            specific_publishers = [p for p in publishers if p.sub == sub]
+            if specific_publishers:
+                return specific_publishers[0]
```

Comment:
Optimize: This might be cleaner with `more_itertools.first_true`, like so:

```suggestion
            if specific_publisher := first_true(publishers, pred=lambda p: p.sub == sub):
                return specific_publisher
```

And in other places that are using this iteration pattern. Definitely worth thinking about other selector patterns from more-itertools, since we have it available for use.

Not a sticking point, but wanted to call out the pattern while I noticed it.